src/app/server/Mdns.cpp

```diff
@@ -471,6 +465,14 @@ void MdnsServer::StartServer(chip::Mdns::CommissioningMode mode)
         ChipLogError(Discovery, "Failed to advertise commissioner: %s", chip::ErrorStr(err));
     }
 #endif // CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY
+
+    // NB: This must be last, so that we announce all of the records we
+    // just added.
+    err = chip::Mdns::ServiceAdvertiser::Instance().Start(&chip::DeviceLayer::InetLayer, chip::Mdns::kMdnsPort);
```

Comment:
The problem here is that `DiscoveryImplPlatform::Start` calls `ChipMdnsStopPublish()` which in many cases removes all the services, so this change will break platform implementations. I don't know how it happened that `DiscoveryImplPlatform` has two methods of opposite-sounding names: `Start` and `StopPublishDevice` which do almost the same :D, but looks like the minimal and platform implementations kinda diverged in their assumptions.

Reply:
Well, the documentation for `ServiceAdvertiser::Start` says:
```
    /// Must be called before Advertise() calls.
```
which is contradicting the behavior of the minimal impl and this proposed change....  It sounds like the platform impls took the API definition at its word?

Reply:
Or possibly that comment was added later after the minimal impl was written, looking at the history....

In any case, we should decide what our API contract here is, both for consumers and implementations.

Reply:
I had noted in #9348 that it was problematic that we don't have an unpublish API that can individually unregister services.

That seems to be contributing greatly to complexity.  The interfaces could be as simple as:

* Advertise(abc) // publish abc service
* DontAdvertise(abc) // stop publishing abc service
* StartServer // start the server and issue unsolicited announcement; don't mutate the list of services
* StopServer // stop the server; don't mutate the list of services

Without this, all of the users of this API are sure to stomp all over each other.

Reply:
These work with platform mDNS because each `Start()` is followed by `AdvertiseOperational` etc. calls. So basically, the assumption was that `Start()` removes all services which are re-added immediately after.

Reply:
I have changed this to simply announce in Advertise(), which will cover the case of adding new records. I'm not exactly happy with it but fixes the regression without having to do overhaul the implementation.